src/NovelRT/Experimental/Graphics/Vulkan/VulkanGraphicsTexture.cpp

```diff
@@ -0,0 +1,348 @@
+
+        VkDevice vulkanDevice = device->GetVulkanDevice();
+        VkDeviceMemory vulkanDeviceMemory = GetBlock()->GetVulkanDeviceMemory();
+
+        void* pDestination = nullptr;
+
+        VkResult result = vkMapMemory(vulkanDevice, vulkanDeviceMemory, GetOffset(), GetSize(), 0, &pDestination);
+
+        if (result != VK_SUCCESS)
+        {
+            throw std::runtime_error("Failed to map VkDeviceMemory! Reason: " + std::to_string(result));
+        }
+
+        return reinterpret_cast<uint8_t*>(pDestination) + rangeOffset;
+    }
+
+    const void* VulkanGraphicsTexture::MapForReadUntyped()
+    {
+        std::shared_ptr<VulkanGraphicsDevice> device = GetAllocator()->GetDevice();
+
+        VkDevice vulkanDevice = device->GetVulkanDevice();
+        VkDeviceMemory vulkanDeviceMemory = GetBlock()->GetVulkanDeviceMemory();
+
+        void* pDestination = nullptr;
+        VkResult mapMemoryResult =
+            vkMapMemory(vulkanDevice, vulkanDeviceMemory, GetOffset(), GetSize(), 0, &pDestination);
+
+        if (mapMemoryResult != VK_SUCCESS)
+        {
+            // TODO: Add real exception.
+            throw std::runtime_error("Failed to map VkMemory to T*! Reason: " + std::to_string(mapMemoryResult));
+        }
+
+        size_t nonCoherentAtomSize =
+            device->GetAdapter()->GetVulkanPhysicalDeviceProperties().limits.nonCoherentAtomSize;
+        size_t offset = GetOffset();
+        size_t size = (GetSize() + nonCoherentAtomSize - 1) & ~(nonCoherentAtomSize - 1);
+
+        VkMappedMemoryRange mappedMemoryRange{};
+        mappedMemoryRange.sType = VK_STRUCTURE_TYPE_MAPPED_MEMORY_RANGE;
+        mappedMemoryRange.memory = vulkanDeviceMemory;
+        mappedMemoryRange.offset = offset;
+        mappedMemoryRange.size = size;
+
+        VkResult invalidateResult = vkInvalidateMappedMemoryRanges(vulkanDevice, 1, &mappedMemoryRange);
+
+        if (invalidateResult != VK_SUCCESS)
+        {
+            // TODO: Add real exception.
+            throw std::runtime_error("Failed to invalidate memory ranges to prevent writing! Reason: " +
+                                     std::to_string(invalidateResult));
+        }
+
+        return pDestination;
+    }
+
+    const void* VulkanGraphicsTexture::MapForReadUntyped(size_t readRangeOffset, size_t readRangeLength)
+    {
+        std::shared_ptr<VulkanGraphicsDevice> device = GetAllocator()->GetDevice();
+
+        VkDevice vulkanDevice = device->GetVulkanDevice();
+        VkDeviceMemory vulkanDeviceMemory = GetBlock()->GetVulkanDeviceMemory();
+
+        void* pDestination = nullptr;
+        VkResult mapMemoryResult =
+            vkMapMemory(vulkanDevice, vulkanDeviceMemory, GetOffset(), GetSize(), 0, &pDestination);
+
+        if (mapMemoryResult != VK_SUCCESS)
+        {
+            // TODO: Add real exception.
+            throw std::runtime_error("Failed to map VkMemory to T*! Reason: " + std::to_string(mapMemoryResult));
```

Comment:
Add a real exception, ye?

Reply:
This will probably be another design ticket or part of an existing one that I mentioned previously.